File: cpp/PhotonLibOS/skynet.cpp
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation: ../TooManyCooks/skynet.cpp
//
// Each node forks 10 children until kDepthMax levels deep (10^kDepthMax =
// 100M leaf tasks) and sums their results. Each child is a
// bench::thread (created on the current vcpu and migrated to a
// work-pool vcpu) joined by the parent, so this measures Photon's thread
// spawn/join throughput under deep nested fan-out.

#include "memusage.hpp"
#include "photon_bench.hpp"

#include <chrono>
#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

static constexpr std::uint64_t kDepthMax = 8;
static const std::size_t iter_count = 1;

static std::uint64_t skynet_one(std::uint64_t base, std::uint64_t depth) {
  if (depth == kDepthMax) {
    return base;
  }
  std::uint64_t offset = 1;
  for (std::uint64_t i = 0; i < kDepthMax - depth - 1; ++i) {
    offset *= 10;
  }

  std::uint64_t results[10];
  std::vector<bench::thread> children;
  children.reserve(10);
  for (std::uint64_t i = 0; i < 10; ++i) {
    const std::uint64_t childBase = base + offset * i;
    std::uint64_t* slot = &results[i];
    children.emplace_back([slot, childBase, depth] {
      *slot = skynet_one(childBase, depth + 1);
    });
  }

  std::uint64_t count = 0;
  for (std::uint64_t i = 0; i < 10; ++i) {
    children[i].join();
    count += results[i];
  }
  return count;
}

static void skynet(std::uint64_t expected) {
  const std::uint64_t count = skynet_one(0, 0);
  if (count != expected) {
    std::printf("ERROR: wrong result - %" PRIu64 "\n", count);
  }
}

int main(int argc, char* argv[]) {
  std::size_t thread_count = bench::default_thread_count();
  if (argc > 1) {
    thread_count = static_cast<std::size_t>(atoi(argv[1]));
  }

  std::uint64_t leaves = 1;
  for (std::uint64_t i = 0; i < kDepthMax; ++i) {
    leaves *= 10;
  }
  const std::uint64_t expected = (leaves - 1) * leaves / 2;

  std::printf("threads: %zu\n", thread_count);

  bench::quiet_logs();
  bench::use_pooled_stacks();

  bench::run_in_pool(thread_count, [expected] {
    skynet(expected); // warmup

    std::printf("runs:\n");
    auto startTime = std::chrono::high_resolution_clock::now();
    for (std::size_t i = 0; i < iter_count; ++i) {
      skynet(expected);
    }
    auto endTime = std::chrono::high_resolution_clock::now();
    auto totalTimeUs = std::chrono::duration_cast<std::chrono::microseconds>(
      endTime - startTime
    );
    std::printf("  - iteration_count: %zu\n", iter_count);
    std::printf("    duration: %" PRIu64 " us\n",
                static_cast<std::uint64_t>(totalTimeUs.count()));
    std::printf("    max_rss: %ld KiB\n", peak_memory_usage());
  });
}

File: rust/bevy_tasks/src/bin/skynet.rs
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/skynet.cpp
//
// Each node forks 10 children until DEPTH_MAX levels deep (10^DEPTH_MAX = 100M
// leaf tasks) and sums their results. The 10-way fork is an iterator over 0..10
// that spawns each child on the pool and collects the `Task` handles; awaiting
// them in order is the join. This measures bevy_tasks' task spawn/join
// throughput under deep nested fan-out.

use bevy_tasks::{block_on, TaskPool};
use bevy_tasks_bench::{default_threads, init_pool, peak_memory_usage, BoxFut};
use std::time::Instant;

const DEPTH_MAX: u64 = 8;
const ITER_COUNT: usize = 1;

fn skynet_one(pool: &'static TaskPool, base: u64, depth: u64) -> BoxFut<u64> {
    Box::pin(async move {
        if depth == DEPTH_MAX {
            return base;
        }
        let mut offset = 1u64;
        for _ in 0..(DEPTH_MAX - depth - 1) {
            offset *= 10;
        }

        let children: Vec<_> = (0..10u64)
            .map(|i| pool.spawn(skynet_one(pool, base + offset * i, depth + 1)))
            .collect();

        let mut count = 0u64;
        for c in children {
            count += c.await;
        }
        count
    })
}

fn skynet(pool: &'static TaskPool, expected: u64) {
    let count = block_on(pool.spawn(skynet_one(pool, 0, 0)));
    if count != expected {
        println!("ERROR: wrong result - {}", count);
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let thread_count = if args.len() > 1 {
        args[1].parse().expect("thread count")
    } else {
        default_threads()
    };

    let leaves = 10u64.pow(DEPTH_MAX as u32);
    let expected = (leaves - 1) * leaves / 2;

    let pool = init_pool(thread_count);
    println!("threads: {}", thread_count);

    skynet(pool, expected); // warmup

    let start = Instant::now();
    for _ in 0..ITER_COUNT {
        skynet(pool, expected);
    }
    let dur = start.elapsed();

    println!("runs:");
    println!("  - iteration_count: {}", ITER_COUNT);
    println!("    duration: {} us", dur.as_micros());
    println!("    max_rss: {} KiB", peak_memory_usage());
}
